### ISSUE TITLE
feat(forge): Cheatcode to receive public key of private key

### ISSUE THREAD

**Issue by pmerkleplant** [foundry-rs/foundry#4790]
### Component

Forge

### Describe the feature you would like

Forge provides a cheatcode to receive the address of a private key, see `forge-std/Vm.sol::addr(uint)(address)`.

It would be great if we could get a cheatcode to receive the public key too. The public key should be returned in affine coordinates, e.g. `vm.pubKey(uint privkey)(uint xCoordinate, unit yCoordinate)`.

To receive the public key of a private key, you currently either have to implement an `ecMul` function in solidity, quickly leading to out-of-gas errors or use `ffi`, leading to big overhead and long-running testsuites.

### Additional context

Issue initially opened in `forge-std` repo: https://github.com/foundry-rs/forge-std/issues/356

**Comment by pmerkleplant** [comment]
This looks great!

There are two more functions I can think of that may benefit from being implemented for such a `Wallet` type:
```solidity
function sign(uint256 privateKey, bytes32 digest) external pure returns (uint8 v, bytes32 r, bytes32 s);
function getNonce(address account) external view returns (uint64 nonce);

// For Wallet:
function sign(Wallet calldata wallet, bytes32 digest) external pure returns (uint8, bytes32, bytes32);
function getNonce(Wallet calldata wallet) external pure returns (unit);
```

This is probably more of an issue for `forge-std` though. `StdWallet.sol` sounds intriguing ;)

**Comment by mds1** [comment]
> This is probably more of an issue for `forge-std` though. `StdWallet.sol` sounds intriguing ;)

Actually per https://github.com/foundry-rs/foundry/issues/3782 we'll end up upstreaming most of those things, and this does feel like a good simple one to upstream from the start. Those extra cheats/overloads make sense to me though

cc'ing @jameswenzel since you might be interested in this discussion per https://github.com/foundry-rs/forge-std/pull/331

**Comment by drewstone** [comment]
I guess the hacky way to get the pubkey would be to sign something and recover the pubkey with ecrecover. Though, I'm looking forward to this feature!

**Comment by pmerkleplant** [comment]
> recover the pubkey with ecrecover.

That's the issue, `ecrecover` does not return the signer's public key but its Ethereum address, see [here](https://www.evm.codes/precompiled#0x01?fork=shanghai).

**Comment by drewstone** [comment]
Gotcha ok, mis-spoke.

**Comment by Evalir** [comment]
Closed with #5332
